enterprise/server/githubapp/githubapp.go

```diff
@@ -1342,3 +1344,142 @@ func (a *GitHubApp) CreateGithubRef(ctx context.Context, req *ghpb.CreateGithubR
 
 	return &ghpb.CreateGithubRefResponse{}, nil
 }
+
+func (a *GitHubApp) GetGithubPullRequest(ctx context.Context, req *ghpb.GetGithubPullRequestRequest) (*ghpb.GetGithubPullRequestResponse, error) {
+	client, err := a.getGithubClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+	incoming, outgoing, user, err := a.getIncomingAndOutgoingPRs(ctx, client)
+	if err != nil {
+		return nil, err
+	}
+	allIssues := append(outgoing.Issues, incoming.Issues...)
+	prs, err := a.populatePRMetadata(ctx, client, allIssues, user)
+	if err != nil {
+		return nil, err
+	}
+	resp := &ghpb.GetGithubPullRequestResponse{
+		Outgoing: []*ghpb.PullRequest{},
+		Incoming: []*ghpb.PullRequest{},
+	}
+	for _, i := range outgoing.Issues {
+		resp.Outgoing = append(resp.Outgoing, prs[i.GetNumber()])
+	}
+	for _, i := range incoming.Issues {
+		resp.Incoming = append(resp.Incoming, prs[i.GetNumber()])
+	}
+	return resp, nil
+}
+
+func (a *GitHubApp) getIncomingAndOutgoingPRs(ctx context.Context, client *github.Client) (*github.IssuesSearchResult, *github.IssuesSearchResult, *github.User, error) {
+	var incoming *github.IssuesSearchResult
+	var outgoing *github.IssuesSearchResult
+	var user *github.User
+	searchGroup, searchContext := errgroup.WithContext(ctx)
+	searchGroup.Go(func() error {
+		r, _, err := client.Search.Issues(searchContext, "is:open is:pr user-review-requested:@me archived:false draft:false", &github.SearchOptions{ListOptions: github.ListOptions{PerPage: 100}})
+		if err != nil {
+			return err
+		}
+		incoming = r
+		return nil
+	})
+	searchGroup.Go(func() error {
+		r, _, err := client.Search.Issues(searchContext, "is:open is:pr author:@me -review:none archived:false draft:false", &github.SearchOptions{ListOptions: github.ListOptions{PerPage: 100}})
+		if err != nil {
+			return err
+		}
+		outgoing = r
+		return nil
+	})
+	searchGroup.Go(func() error {
+		u, _, err := client.Users.Get(ctx, "")
+		if err != nil {
+			return err
+		}
+		user = u
+		return nil
+	})
+	err := searchGroup.Wait()
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	return incoming, outgoing, user, nil
+}
+
+func (a *GitHubApp) populatePRMetadata(ctx context.Context, client *github.Client, prIssues []*github.Issue, currentUser *github.User) (map[int]*ghpb.PullRequest, error) {
+	prsMu := &sync.Mutex{}
+	prs := make(map[int]*ghpb.PullRequest, len(prIssues))
+	metadataGroup, metadataContext := errgroup.WithContext(ctx)
+	for _, i := range prIssues {
+		i := i
+		prsMu.Lock()
+		prs[i.GetNumber()] = issueToPullRequestProto(i)
+		prsMu.Unlock()
+		urlParts := strings.Split(*i.URL, "/")
+
+		metadataGroup.Go(func() error {
+			pr, _, err := client.PullRequests.Get(metadataContext, urlParts[4], urlParts[5], i.GetNumber())
+			if err != nil {
+				return err
+			}
+			prsMu.Lock()
+			p := prs[i.GetNumber()]
```

Comment:
once you've got a hold of p, you can release the lock and edit p freely below (assuming other threads don't simultaneously edit p which i don't think they do)

Reply:
ugh i lied i see in your other thread you are mutating p.Reviews

Reply:
honestly doesn't matter for shit because you're prolly fetching a few hundred things or whatever right

Reply:
Yeah, I thought about having individual mutexes - but felt it was easier to just guard it all with one since there should be max ~200 of these.